File: bindings/python/causalontology/signing.py
"""Record-level signing and verification (spec/provenance.md).

The signature is computed over the record's canonical identity-bearing bytes
(the RFC 8785 form with id and signature removed - exactly the bytes that are
hashed for the record's identifier), so verification needs nothing but the
record itself. Ed25519 is deterministic (RFC 8032): re-signing the same record
with the same key yields the same signature, so re-submission is idempotent.
"""

from . import ed25519
from .canonical import canonicalize, identify, infer_kind


def keypair_from_seed(seed32):
    """(secret, 'ed25519:<hex>') from a 32-byte seed."""
    public = ed25519.secret_to_public(seed32)
    return seed32, "ed25519:" + public.hex()


def sign_record(record, secret, kind=None):
    """Return the record completed with its id and Ed25519 signature."""
    kind = kind or infer_kind(record)
    body = dict(record)
    body.pop("signature", None)
    message = canonicalize(body, kind)
    signature = ed25519.sign(secret, message).hex()
    out = dict(body)
    out["id"] = identify(body, kind)
    out["signature"] = signature
    return out


def _signer_key_hex(record, kind):
    if kind == "succession":
        field = "predecessor"  # a succession is signed by the predecessor key
    else:
        field = "source"
    value = record.get(field, "")
    if not value.startswith("ed25519:"):
        return None
    return value.split(":", 1)[1]


def verify_record(record, kind=None):
    """True iff the record's signature verifies against its own key field."""
    kind = kind or infer_kind(record)
    sig_hex = record.get("signature")
    key_hex = _signer_key_hex(record, kind)
    if not sig_hex or not key_hex:
        return False
    try:
        public = bytes.fromhex(key_hex)
        signature = bytes.fromhex(sig_hex)
    except ValueError:
        return False
    body = dict(record)
    body.pop("signature", None)
    message = canonicalize(body, kind)
    return ed25519.verify(public, message, signature)

File: bindings/csharp/Causalontology/Signing.cs
// Record-level signing and verification (spec/provenance.md).
//
// The signature is computed over the record's canonical identity-bearing
// bytes (the RFC 8785 form with id and signature removed - exactly the
// bytes that are hashed for the record's identifier), so verification
// needs nothing but the record itself. Ed25519 is deterministic
// (RFC 8032): re-signing the same record with the same key yields the
// same signature, so re-submission is idempotent.

namespace Causalontology;

public static class Signing
{
    /// <summary>(secret, "ed25519:&lt;hex&gt;") from a 32-byte seed.</summary>
    public static (byte[] Secret, string PublicId) KeypairFromSeed(byte[] seed32)
    {
        var publicKey = Ed25519.SecretToPublic(seed32);
        return (seed32,
                "ed25519:" + Convert.ToHexString(publicKey).ToLowerInvariant());
    }

    /// <summary>Return the record completed with its id and Ed25519 signature.</summary>
    public static JsonMap SignRecord(JsonMap record, byte[] secret,
                                     string? kind = null)
    {
        kind ??= Canonical.InferKind(record);
        var body = record.Copy();
        body.Remove("signature");
        var message = Canonical.Canonicalize(body, kind);
        var signature = Convert.ToHexString(Ed25519.Sign(secret, message))
            .ToLowerInvariant();
        var output = body.Copy();
        output["id"] = Canonical.Identify(body, kind);
        output["signature"] = signature;
        return output;
    }

    private static string? SignerKeyHex(JsonMap record, string kind)
    {
        // a succession is signed by the predecessor key
        var field = kind == "succession" ? "predecessor" : "source";
        var value = record.GetString(field) ?? "";
        if (!value.StartsWith("ed25519:", StringComparison.Ordinal))
            return null;
        return value.Split(':', 2)[1];
    }

    /// <summary>True iff the record's signature verifies against its own key field.</summary>
    public static bool VerifyRecord(JsonMap record, string? kind = null)
    {
        kind ??= Canonical.InferKind(record);
        var sigHex = record.GetString("signature");
        var keyHex = SignerKeyHex(record, kind);
        if (string.IsNullOrEmpty(sigHex) || string.IsNullOrEmpty(keyHex))
            return false;
        byte[] publicKey, signature;
        try
        {
            publicKey = Convert.FromHexString(keyHex);
            signature = Convert.FromHexString(sigHex);
        }
        catch (FormatException)
        {
            return false;
        }
        var body = record.Copy();
        body.Remove("signature");
        var message = Canonical.Canonicalize(body, kind);
        return Ed25519.Verify(publicKey, message, signature);
    }
}
